### PR TITLE
Milestone #6 Friendships v2

### PR DESCRIPTION
**Pull request by mricanho**
In this milestone, we implement friendships with 2 rows for mutual friendship.

### CODE REVIEW

**Review by umairarshadbutt — CHANGES_REQUESTED** [review]
# Status: REQUIRED CHANGES :warning: :construction: 

Hi @mricanho and @harshdeepkanhai 
Such a remarkable project you made :+1: ! Awesome README :100: !
We, however, do have a few changes to carry out before we proceed (see review comment).

##

 For this project, we have to make sure nothing is messed up. Kindly compare the screenshots below 👇 and work on the suggested changes.
- [x] Please implement friendships with 2 rows for mutual friendship ( for understanding read[this artcile](https://explainextended.com/2009/03/07/selecting-friends) and [stackoverflow discussion](https://stackoverflow.com/questions/4219979/sql-best-practice-for-a-friendship-table))
##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

**Comment by mricanho** [comment]
> # Status: REQUIRED CHANGES  
> Hi @mricanho and @harshdeepkanhai
> Such a remarkable project you made! Awesome README!
> We, however, do have a few changes to carry out before we proceed (see review comment).
> 
> For this project, we have to make sure nothing is messed up. Kindly compare the screenshots below and work on the suggested changes.
> 
> * [x]  Please implement friendships with 2 rows for mutual friendship ( for understanding read[this artcile](https://explainextended.com/2009/03/07/selecting-friends) and [stackoverflow discussion](https://stackoverflow.com/questions/4219979/sql-best-practice-for-a-friendship-table))
> 
> * **When you have made the specified changes, please submit another code review. Thank you!**
> 
> **If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
> HAPPY CODING 
> **[Umair Arshad](https://umairarshad.me/)**
> [Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

HI! 

I already implement the 2 rows since the last milestone, to work with two rows I build the update method of the friendships controller and with the help of another code in user helpers and friendship helper, generates a new record that the status gets changed to true, the code is in:

```
#controllers/friendships_controller.rb

     def update
       @friendship = Friendship.find(params[:id])
       @friendship.update_attribute(:status, true)
       @friendship_two = Friendship.new(user_id: @friendship.friend_id, friend_id: @friendship.user_id, status: true)
       @friendship_two.save
       flash[:notice] = 'Friendship has been accepted'
       friendship_redirect
     end

Greetings!!!

**Review by voscarmv — CHANGES_REQUESTED** [review]
# Impressive work, team. 💻 👀 ✨ 

## However, there are some changes required 👀 🧰 🔨 🔧 
I think we might be missing something in the way you've tried to implement Friendships V2.0.

In your response to the previous TSE you post a snippet of code from your user_helper.rb file. In this piece of code, you seem to be pieceing together a complete list of friends IDs from multiple map filters. This is not the intended way of working in friendships V2.0

Please refer to this example, to better understand how we are expected to use double friendship rows in order to query complete lists of friends without having to piece them together from multiple queries or maps:

https://github.com/voscarmv/ror-social-scaffold/pull/3/commits/bb355e85c086ca345a297ea7bced42406af35b4b#diff-3ce1e0cbd5f25d08bd033ba490fa5e87aa43e1abd54e1c9b017d2a4f06cfa8c1

With double friendship rows, we should be able to run a single query, and obtain a full list of friends, with something similar to this:

https://github.com/voscarmv/ror-social-scaffold/pull/3/commits/06896ca182ff2d4905d152f712a20bc30c325c65

If you have difficulty understanding these examples, please take some time to review the suggested articles by the previous TSE. If you still experience trouble, feel free to contact me on slack. My name is Oscar Mier.


### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Miguel Ricaño Romero & Kanhai Kanhai ** 
Good job with the improvement for this project :muscle:  :muscle:
It meets most of the requirements :clap: :clap: :clap:  
There are just a few details to improve before you can go ahead:

- [x] Don't overuse rubocop disable
- [x] Add integration tests to your project
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

File: app/helpers/friendships_helper.rb [hunk 1]
```diff
@@ -1,4 +1,3 @@
-# rubocop:disable Lint/EmptyConditionalBody
 module FriendshipsHelper
   def verify_friendship(friend_id)
     Friendship.where('(user_id = ? and friend_id = ?) OR (user_id = ? and friend_id = ?)',
```
Comment (acushlakoncept) [inline app/helpers/friendships_helper.rb:3]: [OPTIONAL] You should note methods like this should be better done in your model. 
There should be NO application logic in the views nor helpers (no `Post.new`, `Comment.create`, `current_user.friends.find_by(...)`...). Everything connected with data processing should be encapsulated in models and used in controllers.

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello Miguel Ricaño Romero & Kanhai Kanhai, you've got an impressive project here.

I love the way you used the `content_tag in your helper` and the notification icon. nicely done. 👍  👏 👏  👏  

#### You may have broken a few things while fixing some requested changes.
Let's fix it and we'll be done.

- [x] Your users page is throwing an error. 
Screen Shot 2021-02-14 at 4 54 11 AM.  There is no need to call the User model from there since you already did that in your users controller `@users`. Refactoring your helper method to accept that users will be more appropriate.
```
  def all_users(users)
    users.reject { |x| x.id == current_user.id }
  end
  ```
  and in your users index view page you can call `all_users(@users).each do |user|`

- [x] When a friendship request has been accepted, what did you expect to `yield` here. If nothing, just take out the line and your code should be fine.
```     
 ...   
     elsif friendship.status
      yield
     ...
```
Screen Shot 2021-02-14 at 5 18 32 AM

- [x] I noticed your PR do not have a linter check, kindly fix this too.

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or drop a comment here, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by voscarmv — APPROVED** [review]
# Project status: Approved ✨ 😎👍 💯 ✔️
### You may merge this branch and move forward ⛙🛣

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Review by enaburekhan — COMMENTED** [review]
# Status: Project Approved:heavy_check_mark:
**Hi team!
Nice work:clap::clap:. I really like the Greeting, "Hallo Eric".  Keep up the good work.
Move on with your task:walking:.**

Kindly drop your questions here and ping me on slack @ Eric Enaburekhan to respond.
Have a great day:blush:
Eric Enaburekhan

**Pull request merged**